clang/lib/Driver/Driver.cpp

```diff
@@ -4750,9 +4750,11 @@ Action *Driver::ConstructPhaseAction(
       return C.MakeAction<BackendJobAction>(Input, Output);
     }
     if (Args.hasArg(options::OPT_fsycl_device_only)) {
+      types::ID Output =
+          Args.hasArg(options::OPT_S) ? types::TY_LLVM_IR : types::TY_LLVM_BC;
       if (Args.hasFlag(options::OPT_fsycl_use_bitcode,
                        options::OPT_fno_sycl_use_bitcode, true))
-        return C.MakeAction<BackendJobAction>(Input, types::TY_LLVM_BC);
+        return C.MakeAction<BackendJobAction>(Input, Output);
```

Comment:
Such a renaming would be more readable:
```suggestion
        return C.MakeAction<BackendJobAction>(Input, OutputType);
```